File: src/skilltree/model/Legend.js
// The color legend (F6 — Custom color legend): a tree's ordered list of kinds,
// naming and describing the hues its steps are painted in. A pure domain module —
// every op takes a legend and returns a NEW one, so callers stay immutable and the
// persistence layer + the render always read a fresh value. No I/O lives here.
//
//   kind:   { id, hue, label, description }  (+ derived `count` from withCounts)
//   legend: an ordered array of kinds; hue unique, at most six (the palette).
//
// A node's `color` field IS its kind — it holds the hue. The legend names and
// orders those hues; there is no separate node→kind link. Legend order is
// generation priority: the first kind is the fallback (see F6 §data model).

import { NODE_COLOR_NAMES, DEFAULT_NODE_COLOR } from '../theme.js';

const LABEL_MAX = 24; // one or two words, sentence-case (F6 §data model)
const DESCRIPTION_MAX = 80; // a plain-text sorting brief, optional

// A new/empty tree is born with these three, in order — never all six (F6 §2).
export const DEFAULT_KINDS = [
  { id: 'build', hue: 'terracotta', label: 'Build', description: 'Things you make' },
  { id: 'learn', hue: 'olive', label: 'Learn', description: 'Things you figure out' },
  { id: 'milestone', hue: 'gold', label: 'Milestone', description: 'Moments that matter' },
];

// The tree's legend: saved kinds when we have them (reconciled so every in-use hue
// still has an entry), else derived from the hues actually in use, else the defaults.
export function deriveLegend(nodes, saved) {
  const used = inUseHues(nodes);
  if (saved) {
    const kinds = saved.map((kind) => ({
      id: kind.id,
      hue: kind.hue,
      label: kind.label ?? '',
      description: kind.description ?? '',
    }));
    const present = new Set(kinds.map((kind) => kind.hue));
    const appended = used.filter((hue) => !present.has(hue)).map((hue) => unlabeledKind(hue));
    return [...kinds, ...appended];
  }
  if (nodes.length === 0) return DEFAULT_KINDS;
  return used.map((hue) => unlabeledKind(hue));
}

// Each kind annotated with how many nodes currently wear its hue.
export function withCounts(legend, nodes) {
  const tally = new Map();
  for (const node of nodes) {
    const hue = node.color ?? DEFAULT_NODE_COLOR;
    tally.set(hue, (tally.get(hue) ?? 0) + 1);
  }
  return legend.map((kind) => ({ ...kind, count: tally.get(kind.hue) ?? 0 }));
}

// How many kinds are actually worn by a node — the key appears at two (F6 §4).
export function inUseCount(legend, nodes) {
  return withCounts(legend, nodes).filter((kind) => kind.count > 0).length;
}

// The next unused hue in palette order, or null once all six are taken.
export function freeHue(legend) {
  const taken = new Set(legend.map((kind) => kind.hue));
  return NODE_COLOR_NAMES.find((hue) => !taken.has(hue)) ?? null;
}

export function renameKind(legend, id, label) {
  const clean = label.slice(0, LABEL_MAX);
  return legend.map((kind) => (kind.id === id ? { ...kind, label: clean } : kind));
}

export function describeKind(legend, id, description) {
  const clean = description.slice(0, DESCRIPTION_MAX);
  return legend.map((kind) => (kind.id === id ? { ...kind, description: clean } : kind));
}

// Append an unlabeled kind for a free hue (count 0 until steps use it). A no-op if
// the hue is already taken or the palette is full (hue null). `id` is minted locally
// for a fresh add, or passed through when applying an authoritative AddKind op whose
// id the server (or a collaborator) already chose.
export function addKind(legend, hue, id) {
  if (!hue) return legend;
  if (legend.some((kind) => kind.hue === hue)) return legend;
  return [...legend, unlabeledKind(hue, id)];
}

// Drop a kind from the legend. Callers only remove kinds no node wears (count 0);
// the count guard lives at the call site, which knows the current nodes.
export function removeKind(legend, id) {
  return legend.filter((kind) => kind.id !== id);
}

// Reorder the legend to match `order` (a list of kind ids). Ids absent from `order`
// keep their relative position at the end. Legend order is generation priority.
export function reorderKinds(legend, order) {
  const byId = new Map(legend.map((kind) => [kind.id, kind]));
  const seen = new Set();
  const ordered = [];
  for (const id of order) {
    const kind = byId.get(id);
    if (kind && !seen.has(id)) {
      ordered.push(kind);
      seen.add(id);
    }
  }
  for (const kind of legend) if (!seen.has(kind.id)) ordered.push(kind);
  return ordered;
}

// Swap a kind's hue to a free one, returning the old and new hues so the caller can
// repaint every node of the old hue to the new. A no-op ({ oldHue, newHue } null) if
// the kind is missing or the target hue is already taken by another kind.
export function recolorKind(legend, id, newHue) {
  const target = legend.find((kind) => kind.id === id);
  if (!target || !newHue || legend.some((kind) => kind.hue === newHue)) {
    return { legend, oldHue: null, newHue: null };
  }
  const next = legend.map((kind) => (kind.id === id ? { ...kind, hue: newHue } : kind));
  return { legend: next, oldHue: target.hue, newHue };
}

// The hues a node actually wears, in palette order.
function inUseHues(nodes) {
  const used = new Set(nodes.map((node) => node.color ?? DEFAULT_NODE_COLOR));
  return NODE_COLOR_NAMES.filter((hue) => used.has(hue));
}

function unlabeledKind(hue, id = newId()) {
  return { id, hue, label: '', description: '' };
}

function newId() {
  return crypto.randomUUID?.() ?? `kind-${Date.now()}-${Math.round(Math.random() * 1e6)}`;
}

File: backend/domain/Legend.cpp
#include "domain/Legend.h"

#include <algorithm>

namespace wm {

Legend::Legend(const LegendState& state) {
  for (const KindStateEntry& kind : state.kinds) {
    KindRecord record;
    record.life.addedAt = kind.createdAt;
    record.life.removedAt = kind.deletedAt;
    record.hue = {kind.hue, kind.hueAt};
    record.label = {kind.label, kind.labelAt};
    record.description = {kind.description, kind.descriptionAt};
    record.rank = {kind.rank, kind.rankAt};
    kinds_[kind.id] = std::move(record);
  }
}

Legend::Legend(const std::vector<Kind>& kinds, const Hlc& at) {
  for (std::size_t i = 0; i < kinds.size(); ++i) {
    const Kind& kind = kinds[i];
    KindRecord& record = kinds_[kind.id];
    record.life.add(at);
    record.hue.merge(kind.hue, at);
    record.label.merge(kind.label, at);
    record.description.merge(kind.description, at);
    record.rank.merge(static_cast<double>(i), at);
  }
}

void Legend::join(const LegendState& state) {
  for (const KindStateEntry& kind : state.kinds) {
    KindRecord& record = kinds_[kind.id];
    record.life.add(kind.createdAt);
    record.life.remove(kind.deletedAt);
    record.hue.merge(kind.hue, kind.hueAt);
    record.label.merge(kind.label, kind.labelAt);
    record.description.merge(kind.description, kind.descriptionAt);
    record.rank.merge(kind.rank, kind.rankAt);
  }
}

Legend Legend::seededDefaults(const Hlc& at) {
  return Legend({{KindId{"build"}, NodeColor::terracotta, "Build", "Things you make"},
                 {KindId{"learn"}, NodeColor::olive, "Learn", "Things you figure out"},
                 {KindId{"milestone"}, NodeColor::gold, "Milestone", "Moments that matter"}},
                at);
}

double Legend::nextRank() const {
  double max = -1;
  for (const auto& [id, record] : kinds_) {
    if (record.life.present()) max = std::max(max, record.rank.value);
  }
  return max + 1;
}

void Legend::addKind(const KindId& id, NodeColor hue, const Hlc& at) {
  double rank = nextRank();
  KindRecord& record = kinds_[id];
  record.life.add(at);
  record.hue.merge(hue, at);
  record.rank.merge(rank, at);
}

void Legend::removeKind(const KindId& id, const Hlc& at) {
  kinds_[id].life.remove(at);
}

void Legend::setLabel(const KindId& id, const std::string& label, const Hlc& at) {
  kinds_[id].label.merge(label, at);
}

void Legend::setDescription(const KindId& id, const std::string& description, const Hlc& at) {
  kinds_[id].description.merge(description, at);
}

void Legend::setHue(const KindId& id, NodeColor hue, const Hlc& at) {
  kinds_[id].hue.merge(hue, at);
}

void Legend::reorder(const std::vector<KindId>& order, const Hlc& at) {
  for (std::size_t i = 0; i < order.size(); ++i) {
    auto it = kinds_.find(order[i]);
    if (it != kinds_.end() && it->second.life.present()) it->second.rank.merge(static_cast<double>(i), at);
  }
}

bool Legend::has(const KindId& id) const {
  auto it = kinds_.find(id);
  return it != kinds_.end() && it->second.life.present();
}

bool Legend::empty() const {
  return size() == 0;
}

std::size_t Legend::size() const {
  std::size_t count = 0;
  for (const auto& [id, record] : kinds_) {
    if (record.life.present()) ++count;
  }
  return count;
}

std::optional<NodeColor> Legend::hueOf(const KindId& id) const {
  auto it = kinds_.find(id);
  if (it == kinds_.end() || !it->second.life.present()) return std::nullopt;
  return it->second.hue.value;
}

std::optional<KindId> Legend::ownerOf(NodeColor hue) const {
  for (const auto& [id, record] : kinds_) {
    if (record.life.present() && record.hue.value == hue) return id;
  }
  return std::nullopt;
}

std::optional<Kind> Legend::view(const KindId& id) const {
  auto it = kinds_.find(id);
  if (it == kinds_.end() || !it->second.life.present()) return std::nullopt;
  const KindRecord& record = it->second;
  return Kind{id, record.hue.value, record.label.value, record.description.value};
}

std::vector<Kind> Legend::kinds() const {
  std::vector<std::pair<double, Kind>> ordered;
  for (const auto& [id, record] : kinds_) {
    if (record.life.present())
      ordered.push_back({record.rank.value, Kind{id, record.hue.value, record.label.value, record.description.value}});
  }
  std::sort(ordered.begin(), ordered.end(), [](const auto& a, const auto& b) {
    if (a.first != b.first) return a.first < b.first;
    return a.second.id < b.second.id;
  });

  std::vector<Kind> out;
  for (auto& [rank, kind] : ordered) out.push_back(std::move(kind));
  return out;
}

std::vector<KindId> Legend::orderedIds() const {
  std::vector<KindId> ids;
  for (const Kind& kind : kinds()) ids.push_back(kind.id);
  return ids;
}

LegendState Legend::exportState() const {
  LegendState state;
  for (const auto& [id, record] : kinds_) state.kinds.push_back(*exportKind(id));
  return state;
}

std::optional<KindStateEntry> Legend::exportKind(const KindId& id) const {
  auto it = kinds_.find(id);
  if (it == kinds_.end()) return std::nullopt;
  const KindRecord& record = it->second;
  KindStateEntry entry;
  entry.id = id;
  entry.createdAt = record.life.addedAt;
  entry.deletedAt = record.life.removedAt;
  entry.hue = record.hue.value;
  entry.hueAt = record.hue.stamp;
  entry.label = record.label.value;
  entry.labelAt = record.label.stamp;
  entry.description = record.description.value;
  entry.descriptionAt = record.description.stamp;
  entry.rank = record.rank.value;
  entry.rankAt = record.rank.stamp;
  return entry;
}

}
